source/agent/addons/quic/QuicTransportStream.cc

```diff
@@ -0,0 +1,225 @@
+    unsigned int length = info[1]->Uint32Value();
+    obj->m_stream->Write(buffer, length);
+    info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(QuicTransportStream::addDestination)
+{
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info.Holder());
+    if (info.Length() != 1) {
+        Nan::ThrowTypeError("Invalid argument length for addDestination.");
+        return;
+    }
+    // TODO: Check if info[0] is an Nan wrapped object.
+    auto framePtr = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info[0]->ToObject());
+    // void* ptr = info[0]->ToObject()->GetAlignedPointerFromInternalField(0);
+    // auto framePtr=static_cast<owt_base::FrameDestination*>(ptr);
+    obj->addDataDestination(framePtr);
+    obj->m_isPiped = true;
+}
+
+NAN_METHOD(QuicTransportStream::removeDestination)
+{
+}
+
+v8::Local<v8::Object> QuicTransportStream::newInstance(owt::quic::QuicTransportStreamInterface* stream)
+{
+    Local<Object> streamObject = Nan::NewInstance(Nan::New(QuicTransportStream::s_constructor)).ToLocalChecked();
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(streamObject);
+    obj->m_stream = stream;
+    obj->MaybeReadContentSessionId();
+    return streamObject;
+}
+
+void QuicTransportStream::MaybeReadContentSessionId()
+{
+    if (!m_receivedContentSessionId) {
+        // Match to a content session.
+        if (m_stream->ReadableBytes() > 0 && m_stream->ReadableBytes() < uuidSizeInByte) {
+            ELOG_ERROR("No enough data to get content session ID.");
+            m_stream->Close();
+            return;
+        }
+        uint8_t* data = new uint8_t[uuidSizeInByte];
+        m_stream->Read(data, uuidSizeInByte);
+        m_contentSessionId = std::vector<uint8_t>(data, data + uuidSizeInByte);
+        m_receivedContentSessionId = true;
+        m_asyncOnContentSessionId.data = this;
+        uv_async_send(&m_asyncOnContentSessionId);
+        if (m_stream->ReadableBytes() > 0) {
+            SignalOnData();
+        }
+    }
+}
+
+NAUV_WORK_CB(QuicTransportStream::onData){
+    Nan::HandleScope scope;
+    QuicTransportStream* obj = reinterpret_cast<QuicTransportStream*>(async->data);
+    if (obj == nullptr) {
+        return;
+    }
+    Nan::MaybeLocal<v8::Value> onEvent = Nan::Get(obj->handle(), Nan::New<v8::String>("ondata").ToLocalChecked());
+    if (!onEvent.IsEmpty()) {
+        v8::Local<v8::Value> onEventLocal = onEvent.ToLocalChecked();
+        if (onEventLocal->IsFunction()) {
+            v8::Local<v8::Function> eventCallback = onEventLocal.As<Function>();
+            Nan::AsyncResource* resource = new Nan::AsyncResource(Nan::New<v8::String>("ondata").ToLocalChecked());
+            auto readableBytes = obj->m_stream->ReadableBytes();
+            ELOG_DEBUG("Readable bytes: %d", readableBytes);
+            uint8_t* buffer = new uint8_t[readableBytes]; // Use a shared buffer instead to reduce performance cost on new.
```

Comment:
I don't think it's right to call Nan::NewBuffer with char that's allocated with "new"  My understanding when the object is destroyed, free on that char will be invoked so there's a mismatch here. Can you double check?